crates/bevy_asset/src/asset_changed.rs

```diff
@@ -0,0 +1,468 @@
+///
+/// Unlike `Changed<A>`, this is true whenever the asset for the `A`
+/// in `ResMut<Assets<A>>` changed. For example, when a mesh changed through the
+/// [`Assets<Mesh>::get_mut`] method, `AssetChanged<Mesh>` will iterate over all
+/// entities with the `Handle<Mesh>` for that mesh. Meanwhile, `Changed<Handle<Mesh>>`
+/// will iterate over no entities.
+///
+/// Swapping the actual `A` component is a common pattern. So you
+/// should check for _both_ `AssetChanged<A>` and `Changed<A>` with
+/// `Or<(Changed<A>, AssetChanged<A>)>`.
+///
+/// # Quirks
+///
+/// - Asset changes are registered in the [`AssetEvents`] schedule.
+/// - Removed assets are not detected.
+///
+/// The list of changed assets only gets updated in the
+/// [`AssetEvents`] schedule which runs in `Last`. Therefore, `AssetChanged`
+/// will only pick up asset changes in schedules following `AssetEvents` or the
+/// next frame. Consider adding the system in the `Last` schedule after [`AssetEvents`] if you need
+/// to react without frame delay to asset changes.
+///
+/// # Performance
+///
+/// When at least one `A` is updated, this will
+/// read a hashmap once per entity with a `A` component. The
+/// runtime of the query is proportional to how many entities with a `A`
+/// it matches.
+///
+/// If no `A` asset updated since the last time the system ran, then no lookups occur.
+///
+/// [`AssetEvents`]: crate::AssetEvents
+/// [`Assets<Mesh>::get_mut`]: crate::Assets::get_mut
+pub struct AssetChanged<A: AsAssetId>(PhantomData<A>);
+
+/// [`WorldQuery`] fetch for [`AssetChanged`].
+#[doc(hidden)]
+pub struct AssetChangedFetch<'w, A: AsAssetId> {
+    inner: Option<ReadFetch<'w, A>>,
+    check: AssetChangeCheck<'w, A>,
+}
+
+impl<'w, A: AsAssetId> Clone for AssetChangedFetch<'w, A> {
+    fn clone(&self) -> Self {
+        Self {
+            inner: self.inner,
+            check: self.check,
+        }
+    }
+}
+
+/// [`WorldQuery`] state for [`AssetChanged`].
+#[doc(hidden)]
+pub struct AssetChangedState<A: AsAssetId> {
+    asset_id: ComponentId,
+    resource_id: ComponentId,
+    _asset: PhantomData<fn(A)>,
+}
+
+#[allow(unsafe_code)]
+/// SAFETY: `ROQueryFetch<Self>` is the same as `QueryFetch<Self>`
+unsafe impl<A: AsAssetId> WorldQuery for AssetChanged<A> {
+    type Item<'w> = ();
+    type Fetch<'w> = AssetChangedFetch<'w, A>;
+
+    type State = AssetChangedState<A>;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(_: QueryItem<'wlong, Self>) -> QueryItem<'wshort, Self> {}
+
+    fn shrink_fetch<'wlong: 'wshort, 'wshort>(fetch: Self::Fetch<'wlong>) -> Self::Fetch<'wshort> {
+        fetch
+    }
+
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        state: &Self::State,
+        last_run: Tick,
+        this_run: Tick,
+    ) -> Self::Fetch<'w> {
+        // SAFETY: `AssetChanges` is private and only accessed mutably in the `AssetEvents` schedule
+        let Some(changes) = (unsafe { world.get_resource::<AssetChanges<A::Asset>>() }) else {
```

Comment:
Since you already have the `resource_id` in `state`, it would be slightly more efficient to call `get_resource_by_id`.  I think that would look something like: 

```suggestion
        // SAFETY:
        // - `AssetChanges` is private and only accessed mutably in the `AssetEvents` schedule
        // - `resource_id` was obtained from the type ID of `AssetChanges<A::Asset>`.
        let Some(changes) = (unsafe {
            world
                .get_resource_by_id(state.resource_id)
                .map(|ptr| ptr.deref::<AssetChanges<A::Asset>>())
        }) else {
```